app/assets/javascripts/discourse/app/models/post-stream.js

```diff
@@ -200,49 +220,74 @@ export default RestModel.extend({
   },
 
   cancelFilter() {
-    this.set("summary", false);
-    this.userFilters.clear();
+    this.setProperties({
+      userFilters: [],
+      summary: false,
+      filterRepliesToPostNumber: false,
+      filterUpwardsPostID: false,
+    });
   },
 
-  toggleSummary() {
-    this.userFilters.clear();
-    this.toggleProperty("summary");
-    const opts = {};
-
-    if (!this.summary) {
-      opts.filter = "none";
-    }
-
-    return this.refresh(opts).then(() => {
-      if (this.summary) {
-        this.jumpToSecondVisible();
+  refreshAndJumptoSecondVisible() {
+    return this.refresh({}).then(() => {
+      if (this.posts && this.posts.length > 1) {
+        DiscourseURL.jumpToPost(this.posts[1].get("post_number"));
       }
     });
   },
 
-  jumpToSecondVisible() {
-    const posts = this.posts;
-    if (posts.length > 1) {
-      const secondPostNum = posts[1].get("post_number");
-      DiscourseURL.jumpToPost(secondPostNum);
-    }
+  showSummary() {
+    this.cancelFilter();
+    this.set("summary", true);
+    return this.refreshAndJumptoSecondVisible();
   },
 
   // Filter the stream to a particular user.
-  toggleParticipant(username) {
-    const userFilters = this.userFilters;
-    this.set("summary", false);
+  filterParticipant(username) {
+    this.cancelFilter();
+    this.userFilters.addObject(username);
+    return this.refreshAndJumptoSecondVisible();
+  },
 
-    let jump = false;
-    if (userFilters.includes(username)) {
-      userFilters.removeObject(username);
-    } else {
-      userFilters.addObject(username);
-      jump = true;
-    }
-    return this.refresh().then(() => {
-      if (jump) {
-        this.jumpToSecondVisible();
+  filterReplies(postNumber) {
+    this.cancelFilter();
+    this.set("filterRepliesToPostNumber", postNumber);
+    return this.refresh({ refreshInPlace: true }).then(() => {
+      const element = document.querySelector(`#post_${postNumber}`);
+
+      // order is important, we need to get the offset before triggering a refresh
+      const originalTopOffset = element
+        ? element.getBoundingClientRect().top
+        : null;
+
+      this.appEvents.trigger("post-stream:refresh");
+      DiscourseURL.jumpToPost(postNumber, {
+        originalTopOffset,
+      });
+
+      const replyPostNumbers = this.posts.map((p) => p.post_number);
+      replyPostNumbers.splice(0, 2);
+      schedule("afterRender", () => {
```

Comment:
what happens if this is cancelled before render ? replypostnumbers wont be there anymore ?

Reply:
I think it will be a noop, replypostnumbers would be either empty or have post numbers that don't correspond to elements in the current stream, `highlightPost` handles that.